### ISSUE TITLE
Change Request: Add per-rule types

### ISSUE THREAD

**Issue by thislooksfun** [eslint/eslint#19721]
### ESLint version

9.26.0

### What problem do you want to solve?

ESLint now has first-class types (#18845) which is awesome, but there is still room for improvement, especially regarding auto-complete. This is due to the default type for `Config.rules` just being

https://github.com/eslint/eslint/blob/f60f2764971a33e252be13e560dccf21f554dbf1/lib/types/index.d.ts#L1384C12-L1386

This is functional, and does provide type-safety and autocomplete around the rule severity, but it does not provide safety/autocomplete around the rules used or their options.

My main complaint is that the `rules` object and rule options have no
autocomplete out-of-the-box. This makes it slightly more annoying to try to
write rules quickly as I have to reference external documentation rather than
bringing up autocomplete and reading through the results.

Relatedly, built-in rules' options have no type safety. This means that typos in
rule configs are not caught by TS, even though they totally could be. For
example, currently this snippet will pass type checks:

```ts
export default defineConfig({
  rules: {
    // Typo! "nax" instead of "max"
    complexity: ["error", { nax: 2 }],
    // Typo! Used a string instead of a number
    "max-params": ["error", { max: '2' }]
  },
});
```

### What do you think is the correct solution?

I would like the default definition of `RulesRecord` that ESLint ships with to contain type definitions for all the first-party rules and their options out-of-the-box. I'm no ESLint expert, but as far as I can tell all the information needed to generate these types already exists in machine-readable formats, so unless I'm missing something it should be as "simple" as adding a script / build step that auto-generates the rules definitions.

### Participation

- [x] I am willing to submit a pull request for this change.

### Additional comments

This would also allow for clearer signaling of deprecated options (e.g. the `maximum` form of `max-params`) by adding the `@deprecated` tag to the documentation for that option.

Finally, it would be extra useful if these type definitions would also contain the documentation for the rule / option in question so we can easily read up on what a particular rule or option does without needing to go to the full website documentation. This is a stretch goal however, and would likely require a not-insignificant amount of extra work, so I understand if it is put off / axed. If injecting the full text of the rule's / option's documentation is too much I would settle for a link directly to the documentation.

**Comment by lishaduck** [comment]
For a while, [`eslint-define-config`](https://github.com/eslint-types/eslint-define-config) provided this, but it's since been deprecated (presumably due to maintenance cost, as types have been getting complex since v9).

**Comment by lishaduck** [comment]
I've hacked this together:

```ts
// Based on eslint-define-config, MIT license; @eslint/config-helpers, Apache license

import {
  type Config,
  defineConfig as eslintDefineConfig,
  type ExtendsElement,
} from "@eslint/config-helpers";
import type { Linter } from "eslint";
import type { Rules } from "./rules.ts";

/**
 * Infinite array type.
 */
export type InfiniteArray<T> = T | InfiniteArray<T>[];

type ConfigWithExtendsArray = InfiniteArray<ConfigWithExtends>[];

/**
 * Config with extends. Valid only inside of `defineConfig()`.
 */
export interface ConfigWithExtends extends Linter.Config<Rules> {
  extends?: ExtendsElement[];
}

export const defineConfig: (...args: ConfigWithExtendsArray) => Config[] =
  eslintDefineConfig;
```

```ts
import type { Linter } from "eslint";
import type { EslintRules } from "./builtin/index.ts";

// eslint-disable-next-line @typescript-eslint/no-empty-object-type -- Interface merging.
interface CustomRuleOptions {}

type CustomRules = {
  [TRuleName in keyof CustomRuleOptions]: Linter.RuleSeverityAndOptions<
    CustomRuleOptions[TRuleName]
  >;
};

export interface Rules extends CustomRules, EslintRules, Linter.RulesRecord {}
```

**Comment by ntnyq** [comment]
You can use https://github.com/antfu/eslint-typegen to generate the json-schema based rule options.

**Comment by nzakas** [comment]
We actually do ship type definitions for built-in rules:
https://github.com/eslint/eslint/blob/0f49329b4a7f91714f2cd1e9ce532d32202c47f4/lib/types/rules.d.ts#L97

It just needs to manually be applied in your code.

It would be nice to get this type information flowing out-of-the-box, but we'd want to be able to do this for all rules and not just built-in ones. We'll be moving a lot of the core type information into the `@eslint/core` package and out of this one, so we don't want to implement anything new in this package that would need to be replaced later. 

Moving this to blocked until we have time to think this through.

**Comment by lishaduck** [comment]
> We actually do ship type definitions for built-in rules:

Cool, that's really helpful. Just dropped (most) of eslint-typegen's buggy interface merging in favor of that.
Only thing I found is that curly doesn't support the consistent option. I think 
https://github.com/eslint/eslint/blob/0f49329b4a7f91714f2cd1e9ce532d32202c47f4/lib/types/rules.d.ts#L622

should be:

```ts
["all" | "multi" | "multi-line" | "multi-or-nest", "consistent"?]
```

at least if I'm reading the docs right.

**Comment by lishaduck** [comment]
Until the day when [eslint/rewrite#208](https://github.com/eslint/rewrite/issues/208) shall get resolved, here's my version:

`@deputy-eslint/define-config`:

```ts
import {
  defineConfig as eslintDefineConfig,
  type ExtendsElement,
} from "@eslint/config-helpers";
import type { Linter } from "eslint";
import type { ESLintRules } from "eslint/rules";

export interface RulesRecord extends ESLintRules {}

export type TypedRules = Partial<RulesRecord>;
export type TypedConfig = Linter.Config<RulesRecord>;

export interface ConfigWithExtends extends TypedConfig {
  extends?: ExtendsElement[];
}

export type InfiniteArray<T> = InfiniteArray<T>[] | T;

export type ConfigWithExtendsArray = InfiniteArray<ConfigWithExtends>;

export const defineConfig = (
  ...configs: ConfigWithExtendsArray[]
): TypedConfig[] => eslintDefineConfig(...configs);
```

`@deputy-eslint/typegen`:

```
import fs from "node:fs/promises";

import { flatConfigsToRulesDTS } from "eslint-typegen/core";
import { concat, type Awaitable } from "eslint-flat-config-utils";
import type { Linter } from "eslint";

export async function generate(
  ...configs: Awaitable<Linter.Config | Linter.Config[]>[]
): Promise<void> {
  const config = await concat(...configs);

  let dts = await flatConfigsToRulesDTS(config, {
    includeAugmentation: true,
  });

  dts += `
declare module "@eslint-deputy/define-config" {
  interface RulesRecord extends RuleOptions {}
}
`;

  dts = dts.replaceAll("\ntype", "\n/** @internal */\nexport type");

  await fs.writeFile("src/typegen.d.ts", dts);
}
```

`@eslint-deputy/config/scripts/typegen.ts`:

```ts
#!/usr/bin/env node
import { generate } from "@eslint-deputy/typegen";

import { allConfigs } from "../src/configs/index.ts";

await generate(...allConfigs);
```

**Comment by ntnyq** [comment]
> > We actually do ship type definitions for built-in rules:
> 
> Cool, that's really helpful. Just dropped (most) of eslint-typegen's buggy interface merging in favor of that. Only thing I found is that curly doesn't support the consistent option. I think
> 
> [eslint/lib/types/rules.d.ts](https://github.com/eslint/eslint/blob/0f49329b4a7f91714f2cd1e9ce532d32202c47f4/lib/types/rules.d.ts#L622)
> 
> Line 622 in [0f49329](/eslint/eslint/commit/0f49329b4a7f91714f2cd1e9ce532d32202c47f4)
> 
>  ["all" | "multi" | "multi-line" | "multi-or-nest" | "consistent"] 
> should be:
> 
> ["all" | "multi" | "multi-line" | "multi-or-nest", "consistent"?]
> at least if I'm reading the docs right.

According the json-schema defined 

https://github.com/eslint/eslint/blob/0f49329b4a7f91714f2cd1e9ce532d32202c47f4/lib/rules/curly.js#L30-L56

It should be `["all"] | ["multi" | "multi-line" | "multi-or-nest", "consistent"?]`

**Comment by nzakas** [comment]
If someone could submit a PR to fix that, we'd appreciate it.

**Comment by lishaduck** [comment]
> If someone could submit a PR to fix that, we'd appreciate it.

I just went through and diffed the eslint-typegen output and the eslint/rules output and found 48 incompatibilities.
PR incoming for curly, I'll move the rest of that discussion over there.

**Comment by Shinigami92** [comment]
> For a while, [`eslint-define-config`](https://github.com/eslint-types/eslint-define-config) provided this, but it's since been deprecated (presumably due to maintenance cost, as types have been getting complex since v9).

Hi 👋 I'm the maintainer of this now deprecated package and nope, I deprecated it in favor of eslint's native `defineConfig` and want that the community ships exactly what this issue is asking for 😉

**Comment by lumirlumir** [comment]
Marking this issue as accepted, as there is an ongoing PR addressing it: https://github.com/eslint/eslint/pull/19843
